File: 2019-11-08/trickortreat.py
def trickOrTreat(children, bags):
  if len(bags) >= children:
    if len(children_will_accept_contents_of(bags)) >= children:
        return "Thank you, strange uncle!"
  return "Trick or treat!"

def children_will_accept_contents_of(bags):
  return [ bag for bag in bags if number_of_candies_in(bag) >= 2 and "bomb" not in bag ]

def number_of_candies_in(bag):
  return len([ treat for treat in bag if treat == "candy"])

BAG_OF_ONE_CANDY = ["candy"]
BAG_OF_TWO_CANDIES = ["candy", "candy"]
BAG_OF_TWO_APPLES = ["apple", "apple"]

def shouldSayTrickOrTreat(children, bags):
  assert trickOrTreat(children, bags) == "Trick or treat!"

def shouldSayThankYou(children, bags):
  assert trickOrTreat(children, bags) == "Thank you, strange uncle!"

def test_no_bags():
  shouldSayTrickOrTreat(2, [])

def test_enough_candies_for_everyone():
  shouldSayThankYou(2, [BAG_OF_TWO_CANDIES] * 2)
  shouldSayThankYou(3, [BAG_OF_TWO_CANDIES] * 3)
  shouldSayThankYou(4, [BAG_OF_TWO_CANDIES] * 5)

def test_not_enough_candies_for_everyone():
  shouldSayTrickOrTreat(2, [BAG_OF_TWO_CANDIES] * 1)
  shouldSayTrickOrTreat(3, [BAG_OF_TWO_CANDIES] * 2)

def test_not_enough_candies_per_child():
  shouldSayTrickOrTreat(2, [BAG_OF_ONE_CANDY] * 2)
  shouldSayTrickOrTreat(2, [BAG_OF_TWO_CANDIES, BAG_OF_ONE_CANDY])
  shouldSayTrickOrTreat(2, [BAG_OF_ONE_CANDY] * 3)
  shouldSayTrickOrTreat(3, [BAG_OF_ONE_CANDY] * 3)

def test_ignore_apples():
  shouldSayTrickOrTreat(2, [BAG_OF_TWO_CANDIES, ["candy", "apple"]])
  shouldSayTrickOrTreat(2, [BAG_OF_TWO_CANDIES, BAG_OF_TWO_APPLES])
  shouldSayThankYou(2, [BAG_OF_TWO_CANDIES, BAG_OF_TWO_CANDIES, BAG_OF_TWO_APPLES])

def test_bomb():
  shouldSayTrickOrTreat(2, [BAG_OF_TWO_CANDIES, BAG_OF_TWO_CANDIES + ["bomb"]])

# def test_each_child_has_same_number_of_candies():
#   shouldSayTrickOrTreat(2, [BAG_OF_TWO_CANDIES, BAG_OF_ONE_CANDY * 3])

File: 2020-01-30/TrickOrTreat.hs
import Test.Hspec

import Data.Function ((&))
import Data.List (sort, group)

main = hspec $ do
    describe "The treat or trick kata" $ do
        it "has a canary" $ do
            1 `shouldBe` 1
        it "two children, happy scenario" $ do
            distribute 2 [["Candy", "Candy"]
                         ,["Candy", "Candy"]
                         ]
                `shouldBe` "Thank you, strange uncle!" 
            distribute 3 [["Candy", "Candy"]
                         ,["Candy", "Candy"]
                         ,["Candy", "Candy"]
                         ]
                `shouldBe` "Thank you, strange uncle!" 
            distribute 3 [["Candy", "Candy", "Candy"]
                         ,["Candy", "Candy", "Candy"]
                         ,["Candy", "Candy", "Candy"]
                         ]
                `shouldBe` "Thank you, strange uncle!" 
        it "does not work when there are less bags than children" $ do
            distribute 2 []
                `shouldBe` "Tricks or treats"
            distribute 2 [["Candy", "Candy"]]
                `shouldBe` "Tricks or treats"
            distribute 2 [[]]
                `shouldBe` "Tricks or treats"
        it "empty bags" $ do
            distribute 2 [[], []]
                `shouldBe` "Tricks or treats"
        it "kids gets keep one apple" $ do
            distribute 2 [["Apple"]
                         ,["Apple"]
                         ]
                `shouldBe` "Tricks or treats"
        it "kids gets one apple one candy" $ do
            distribute 2 [["Apple", "Candy"]
                         ,["Apple", "Candy"]
                         ]
                `shouldBe` "Tricks or treats"
        it "kids don't like bombs" $ do
            distribute 3 [["Candy", "Candy", "Candy"]
                         ,["Candy", "Candy", "Candy", "Bomb"] -- EVIL LAUGH
                         ,["Candy", "Candy", "Candy"]
                         ]
                `shouldBe` "Tricks or treats" 
        it "kids should receive the same number of candies" $ do
            distribute 3 [["Candy", "Candy", "Candy"]
                         ,["Candy", "Candy"]
                         ,["Candy", "Candy", "Candy"]
                         ]
                `shouldBe` "Tricks or treats" 

type Treat = String
type Bag = [Treat]
type Answer = String

-- `|' means such that
distribute :: Int -> [Bag] -> Answer
distribute children bags 
    | (compliant bags & map nbCandies & sort & group & map length & foldl max 0) >= children = "Thank you, strange uncle!" 
distribute _ _ = "Tricks or treats"

compliant = filter isCompliant

isCompliant :: Bag -> Bool
isCompliant bag | "Bomb" `elem` bag = False
isCompliant bag = length (filter keepCandies bag) >= 2 

keepCandies :: Treat -> Bool
keepCandies treat = treat == "Candy"    

nbCandies bag =
    bag 
    & filter keepCandies
    & length
